kubernetes/client.go

```diff
@@ -139,6 +145,7 @@ func getConfigForLocalCluster() (*rest.Config, error) {
 			log.Errorf("Error '%v' getting config for local cluster", err.Error())
 			return nil, err
 		}
+		incluster.BearerToken = "" // force use of BearerTokenFile so client can refresh the token
```

Comment:
I don't think you need to set this explicitly here either since it gets set already in `rest.InClusterConfig`: https://github.com/kubernetes/client-go/blob/fb8b7346aacefea5ee2ab2e234afc4451c90c435/rest/config.go#L511-L541.

Reply:
This code now sets BearerTokenFile everywhere that it is known, and it also sets BearerToken to "" everywhere that the BearerTokenFile is known. I did this for two reasons:

1. This makes everything consistent - where a file is known, then set it; and blank the token in those instances. Otherwise we'll have code that will look strange to future devs ("why is the file set here but not here?"). Just always set the bearer file everywhere that it is known. The file is used under the covers by the go-client and the token is loaded at that time (at least, that's how I am seeing it work). If, for some reason, setting the BearerTokenFile is a no-op and that file is not read to obtain the token, that we can look to adjust this (but if that really happens and the file is ignored, that sounds like a bug - which gets me to my second reason why I did this...)
2. I want to ensure that BearerTokenFile is actually being used - if we set both, we actually cannot be assured all underlying code is using the File if it also has the Token. And we'll never know we have a bug unless some one tests a token past its expiration time (and that would only happen when deployed out in the wild since we don't have any tests that literally test an expiring token).

My latest commit has a unit test that now looks at the bearer token and bearer token file fields to make sure they are set for the home cluster and a remote cluster.

Reply:
> I want to ensure that BearerTokenFile is actually being used - if we set both, we actually cannot be assured all underlying code is using the File if it also has the Token.

I think you can legitimately set one but not the other i.e. you can set `BearerToken` but not `BearerTokenFile`. Right now for the remote config, we only override `qps` and `burst` but all other settings come straight from user configuration and I don't want that to change by overriding more fields: https://github.com/kiali/kiali/blob/7eb710750713a2ee70a9a6f6d1252b3da2ac9f4a/kubernetes/client_factory.go#L491-L497

The only thing that needs to change is for the kiali SA client for the home cluster to also set the `BearerTokenFile`. If we were to just use `rest.InClusterConfig` then this would be happening already. For instance, I'm not sure why we copy over the token here instead of using `rest.InClusterConfig`: https://github.com/kiali/kiali/blob/7eb710750713a2ee70a9a6f6d1252b3da2ac9f4a/kubernetes/client_factory.go#L499-L507

Reply:
> I think you can legitimately set one but not the other i.e. you can set `BearerToken` but not `BearerTokenFile`.

Docs imply you can set both, but a non-empty File means the Token will be refreshed (except for when using OAuth2, which is why I wasn't sure if I should leave Token as-is or empty it out.. I'm thinking I'm going to revert that change and just leave them both set.). Here's the docs:

https://pkg.go.dev/k8s.io/client-go@v0.28.3/rest#Config.BearerToken

```go
// Server requires Bearer authentication. This client will not attempt to use
// refresh tokens for an OAuth2 flow.
// TODO: demonstrate an OAuth2 compatible client.
BearerToken [string](https://pkg.go.dev/builtin#string) `datapolicy:"token"`

// Path to a file containing a BearerToken.
// If set, the contents are periodically read.
// The last successfully read value takes precedence over BearerToken.
BearerTokenFile [string](https://pkg.go.dev/builtin#string)
```

Reply:
> For instance, I'm not sure why we copy over the token here instead of using `rest.InClusterConfig`

Yeah, there was a couple things in the code that made me scratch my head wondering why we were doing things. I think changes happened rapidly in here and we just missed a few places.

Reply:
I reverted some of the code I originally pushed. I no longer blank the token. The docs imply that the token will be overridden if File is provided and the File is refreshed. My paranoia just caused me to do more work, so I reverted that. I'll trust the go-client does the right thing.

Reply:
> For instance, I'm not sure why we copy over the token here instead of using `rest.InClusterConfig`:

Well, I found out why we do this. It uses the base config that was passed to the factory, not the InClusterConfig. And it does this because it is mock-friendly (we must pass in mock configs when creating factories in tests - indeed, I see errors in the tests when I changed this code to use InClusterConfig).

See: https://github.com/kiali/kiali/blob/v1.78.0/kubernetes/client_factory.go#L119-L121

```
// newClientFactory allows for specifying the config and expiry duration
// Mock friendly for testing purposes
func newClientFactory(restConfig *rest.Config) (*clientFactory, error) {
```